quickwit/quickwit-common/src/tower/rate_estimator.rs

```diff
@@ -119,59 +124,57 @@ impl RateEstimator for SmaRateEstimator {
     fn update(&mut self, _started_at: Instant, ended_at: Instant, work: u64) {
         let elapsed = ended_at.duration_since(self.inner.anchor).as_secs();
         let bucket_ord = elapsed / self.inner.bucket_period_secs;
-        let bucket_idx = (bucket_ord % self.inner.num_buckets) as usize;
-
-        let current_bucket = &self.inner.buckets[bucket_idx];
-        let current_bucket_ord = current_bucket.bucket_ord();
-
-        if current_bucket_ord > bucket_ord {
-            return;
-        }
-        if current_bucket_ord != bucket_ord {
-            current_bucket.reset(current_bucket_ord, bucket_ord);
-        }
-        current_bucket.update(work);
+        let num_buckets = self.inner.num_buckets;
+        let bucket_idx = bucket_ord % num_buckets;
+        let bucket = &self.inner.buckets[bucket_idx as usize];
+        bucket.increment_work(work, bucket_ord, num_buckets);
     }
 }
 
+/// Rate estimator bucket. The 63 least significant bits of the atomic integer store the amount of
+/// work, while the most significant bit is used to indicate whether the bucket needs to be reset.
+/// The reset bit is also called the "color" of a bucket in an attempt to make the code more
+/// readable. After each complete pass over the buckets, the color is flipped. The color `0`
+/// corresponds to the even passes, while the color `1` corresponds to the odd passes.
 #[derive(Debug, Default)]
 struct Bucket {
-    bucket_ord: AtomicU64,
-    cumulated_work: AtomicU64,
+    bits: AtomicU64,
 }
 
 impl Bucket {
-    fn new(bucket_ord: u64, cumulated_work: u64) -> Self {
-        Self {
-            bucket_ord: AtomicU64::new(bucket_ord),
-            cumulated_work: AtomicU64::new(cumulated_work),
-        }
-    }
+    const COLOR_MASK: u64 = 0x8000000000000000;
 
-    fn bucket_ord(&self) -> u64 {
-        self.bucket_ord.load(Ordering::Acquire)
-    }
+    const WORK_MASK: u64 = 0x7FFFFFFFFFFFFFFF;
 
-    fn cumulated_work(&self) -> u64 {
-        self.cumulated_work.load(Ordering::Relaxed)
+    fn work(&self) -> u64 {
+        self.bits.load(Ordering::Relaxed) & Self::WORK_MASK
     }
 
-    fn update(&self, work: u64) {
-        self.cumulated_work.fetch_add(work, Ordering::Relaxed);
-    }
+    fn increment_work(&self, work: u64, bucket_ord: u64, num_buckets: u64) {
```

Comment:
This should be `self.bits.fetch_add(work & Self::WORK_MASK, Ordering::Relaxed)` but that seems overkill.